### PR TITLE
backend: Move semaphores to a dedicated package

### PR DESCRIPTION
**Pull request by greatroar**
What does this PR change? What problem does it solve?
-----------------------------------------------------

backend.Semaphore has been moved to a dedicated package called internal/sema. I resisted the temptation to call the main type sema.Phore. Also, semaphores are now passed by value to skip a level of indirection when using them.

I was trying some other refactoring of the backend and cmd/restic packages, in order to get a REST API working. I found I got import cycles, one of them involving backend.Semaphore. Since semaphores are a generally useful thing to have, I decided to move them out of the way first.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

No.

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

File: internal/sema/semaphore.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+type wrapReader struct {
+	io.ReadCloser
+	eofSeen bool
+	sem     Semaphore
+	cancel  context.CancelFunc
+}
+
+func (wr *wrapReader) Read(p []byte) (int, error) {
+	if wr.eofSeen { // XXX Why do we do this?
```
Comment (MichaelEischer) [inline internal/sema/semaphore.go:47]: I have no idea. It seems to originate in #978. @fd0 do you remember the reason for remembering whether the read is already closed?

**Review by MichaelEischer — COMMENTED** [review]
>  I found I got import cycles, one of them involving backend.Semaphore. Since semaphores are a generally useful thing to have, I decided to move them out of the way first.

That sounds like you ended up with the backend package importing a backend?

I currently don't see any user for the Semaphore except the backends, so maybe keep them inside `internal/backend` to not clutter `internal`?

**Comment by greatroar** [comment]
> That sounds like you ended up with the backend package importing a backend?

All of them, in fact. I wanted to move parseConfig, open and close from cmd/global.go to backend (or some package inside it), so you can do an API request to open a repo.

> I currently don't see any user for the Semaphore except the backends, so maybe keep them inside internal/backend to not clutter internal?

Done.

**Review by MichaelEischer — APPROVED** [review]
LGTM.

**Pull request merged**